File: Assignment3/q3.py
#q3
'''Printing False means the determinant of given matrix is zero'''
def partial_pivot(ar,b):
    n=len(ar)
    zero=[0 for _ in range(n)]
    for i in range(n):
        if ar[i][i]==0:
            if ar[i]==zero:
                return False
            j=i+1;t=False
            while j<n and not t:
                if ar[j][i]!=0:
                    ar[j],ar[i]=ar[i],ar[j]
                    b[j], b[i] = b[i], b[j]
                    t=True
    return True

def gauss_jordan(arr):
    ar=arr[:]
    n = len(ar)
    b=[[1 if i==j else 0 for j in range(n)] for i in range(n)]
    if(n!=len(b)):
        print("Invalid input")
        return
    for i in range(n):
        a=partial_pivot(ar,b)
        if not a:
            return a
        b[i]=[b[i][t] / ar[i][i] for t in range(n)]
        ar[i] = [ar[i][t] / ar[i][i] for t in range(n)]
        for j in range(n):
            if i==j or ar[j][i]==0:
                continue
            else:
                b[j]= [b[j][t]-b[i][t]*ar[j][i] for t in range(n)]
                ar[j] = [ar[j][t] - ar[i][t] * ar[j][i] for t in range(n)]
    return b

def matrix_mult(a,b):
    axb=[];m=len(b)
    for i in range(m):
        axb.append([])
        for j in range(m):
            sum=0
            for t in range(len(b)):
                sum+=a[i][t]*b[t][j]
            axb[i].append(sum)
    return axb

matrix=[]
with open("q3.txt","r") as f:
    a=f.readlines()
    for i in a:
        x=i.split("\n")[0].split(" ")
        n=len(x)
        matrix.append([float(x[j]) for j in range(n)])
print("Inverse=",C:=gauss_jordan(matrix))
print("Multiplication=",matrix_mult(C,matrix))
#Output
'''
Inverse= [[-3.0, 3.0, -7.0], [1.0, 1.0, 0.0], [1.0, 0.0, 1.0]]
Multiplication= [[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]
'''

File: Assignment3-cpp/q3.cpp
#include<iostream>
using namespace std;
#define LIMIT 3

void swap(float &a,float &b){
	float t=a;a=b;b=t;
}
bool partial_pivot(float ar[][LIMIT],float br[][LIMIT]){
	float zero[]={0,0,0};
	for(int i=0;i<LIMIT;i++){
		if (ar[i][i]==0){
			bool isit=true;
			for(int t=0;t<LIMIT;t++){
				if(zero[t]!=ar[i][t])isit=false; 
			}
			if(isit) return (!isit);
			for(int j=i+1;i<LIMIT;i++){
				if(ar[j][i]!=0){
					for(int t=0;t<LIMIT;t++){
						swap(ar[j][t],ar[i][t]);
						swap(br[j][t],br[i][t]);
					}
					i=LIMIT;
				}
			}
		}
	}
	return true;
}

void gauss_jordan(float a[][LIMIT],float br[][LIMIT]){
	float ar[LIMIT][LIMIT];
	for(int i=0;i<LIMIT;i++){
		for(int j=0;j<LIMIT;j++){ar[i][j]=a[i][j];}
	}
	for(int i=0;i<LIMIT;i++){
		bool c=partial_pivot(ar,br);
		if(!c) {cout<<"WARNING: DETERMINANT IS 0."<<endl;return;}
		float pivot=ar[i][i];
		for(int j=0;j<LIMIT;j++){ar[i][j]/=pivot;br[i][j]/=pivot;}
		for(int j=0;j<LIMIT;j++){
			if(i==j or ar[j][i]==0)continue;
			else{
				float fact=ar[j][i];
				for(int t=0;t<LIMIT;t++){
					ar[j][t]-=fact*ar[i][t];
					br[j][t]-=fact*br[i][t];
				}
			}
		}
	}
}
void multiplication(float ar[][LIMIT],float br[][LIMIT]){
	cout<<"Matrix Multiplication=\n[";
	for(int i=0;i<LIMIT;i++){
		cout<<"[";
		for(int j=0;j<LIMIT;j++){
			float sum=0;
			for(int t=0;t<LIMIT;t++){
				sum+=ar[i][t]*br[t][j];
			}
			cout<<sum<<" ";
		}
		if (i==LIMIT-1)cout<<"]]\n";
		else cout<<"]\n";
	}
}
int main(){
	float ar[LIMIT][LIMIT],
	Inv[][LIMIT]={{1,0,0},{0,1,0},{0,0,1}};
	FILE* f=fopen("q3.txt","r");
	for(int i=0;i<LIMIT;i++){
		for(int j=0;j<LIMIT;j++){
			fscanf(f,"%f",&ar[i][j]);
		}
	}
	fclose(f);
	cout<<"Inverse of given matrix=\n[";
	gauss_jordan(ar,Inv);
	for(int i=0;i<LIMIT;i++){
		cout<<"[";
		for(int j=0;j<LIMIT;j++){
			cout<<Inv[i][j]<<" ";
		}
		if (i==LIMIT-1)cout<<"]]\n";
		else cout<<"]\n";
	}
	
	multiplication(ar,Inv);
}
/*
Output:
Inverse of given matrix=
[[-3 3 -7 ]
[1 1 0 ]
[1 0 1 ]]
Matrix Multiplication=
[[1 0 0 ]
[0 1 0 ]
[0 0 1 ]]
*/